### DISCUSSION TITLE
MR/XR/XRC Connection Details multi-tenant isolation not possible

### DISCUSSION THREAD

**Discussion by jaylevin** [crossplane/crossplane#4227] [General]
### Background/Context
We have multiple tenants using our control plane to deploy infrastructure. Each tenant has their own namespace in our control plane. We want each tenant to have their own Vault KV Engine where their secrets will land. We want to ensure that when the tenant submits a claim, the associated MR, XR, and XRC connection details all land in their designated Vault KV Engine.

My understanding is this is impossible because the `publishConnectionDetailsWithStoreConfigRef` field is defined at the top layer of the `Composition`, and there is no way to patch this value. This results in all connection details for XRs to land in the same KV Engine (since they all share the same `StoreConfig`/`mountPath`).

My understanding is:

1. The XRC's `publishConnectionDetailsTo` `StoreConfig` reference can achieve tenant isolation by using K8s Policies like Datree or OPA. (Force them to only be able to reference the `StoreConfig` for their tenant namespace that is created by platform engineers).
2. The MR's `publishConnectionDetailsTo` `StoreConfig` reference can be constrained using Crossplane built-in P&T logic.
3. The XR's `publishConnectionDetailsWithStoreConfigRef` `StoreConfig` reference cannot be constrained, which results in the connection details for all of our tenants landing in a single KV engine, violating our security requirements.

Are my assumptions correct? Is there something I am missing?

Any help would be greatly appreciated.

**Comment by jeanduplessis** [comment]
@jaylevin, we are trying to keep issues in this repo specific to bugs/feature requests. As this issue is currently more of a discussion/request for clarification, I will convert it to a discussion where the community can further collaborate on it.

**Reply by jaylevin** [reply]
@jeanduplessis I would argue that this is a bug report/feature request.

I would expect Crossplane to support this feature, so I'm not sure if it's a bug (this would be a bug) or if it wasn't considered during design phase (then this would be a feature request).

At the end of the day:
We should be able to isolate the connection details (at the XR, MR, and XRC) on a per tenant basis. This means that the XR connection details created as a result of a claim should be able to be isolated to a tenant (claim namespace) specific Key Vault Engine. Since the `publishConnectionDetailsWithStoreConfigRef` cannot be patched, the XR connection details for every tenant (claim namespace) end up landing in the same Vault key engine (due to using the same `StoreConfig` reference and associated `mountPath`).
